src/GovUk.Education.ExploreEducationStatistics.Publisher/Services/PermalinkMigrationService.cs

```diff
@@ -0,0 +1,127 @@
+
+        do
+        {
+            var pages = blobContainer
+                .GetBlobsAsync(BlobTraits.None, prefix: null)
+                .AsPages(continuationToken);
+
+            await foreach (var page in pages)
+            {
+                var messages = page.Values.Select(blobItem =>
+                {
+                    var name = blobItem.Name;
+                    var permalinkId = Guid.Parse(name);
+                    return new PermalinkMigrationMessage(permalinkId);
+                }).ToList();
+
+                await _storageQueueService.AddMessages(PermalinkMigrationQueue, messages);
+
+                continuationToken = page.ContinuationToken;
+            }
+        } while (continuationToken != string.Empty);
+    }
+
+    public async Task<Permalink> AddPermalinkToDbFromStorage(Guid permalinkId)
+    {
+        var permalink = await GetPermalinkFromStorage(permalinkId);
+
+        if (!await _contentDbContext.Permalinks.AnyAsync(p => p.Id == permalinkId))
+        {
+            _contentDbContext.Permalinks.Add(permalink);
+            await _contentDbContext.SaveChangesAsync();
+        }
+
+        return permalink;
+    }
+
+    private async Task<Permalink> GetPermalinkFromStorage(Guid permalinkId)
+    {
+        var blobContainer = _blobServiceClient.GetBlobContainerClient(Permalinks.Name);
+        var blob = blobContainer.GetBlobClient(permalinkId.ToString());
+
+        if (!await blob.ExistsAsync())
+        {
+            throw new InvalidOperationException($"Blob not found for permalink {permalinkId}");
+        }
+
+        await using var stream = await blob.OpenReadAsync();
+        using var streamReader = new StreamReader(stream);
+        using var jsonReader = new JsonTextReader(streamReader);
+
+        var jsonObject = await JToken.ReadFromAsync(jsonReader);
+
+        var id = jsonObject.Value<string>("Id");
+        var created = jsonObject.Value<DateTime>("Created");
+
+        var subjectId = jsonObject.SelectToken("$.Query.SubjectId")?.ToObject<string>();
+        if (subjectId == null)
+        {
+            throw new InvalidOperationException("Permalink found with no SubjectId");
+        }
+
+        var publicationName = jsonObject.SelectToken("$.FullTable.SubjectMeta.PublicationName")?.ToObject<string>();
+        if (publicationName == null)
+        {
+            throw new InvalidOperationException("Permalink found with no PublicationName");
+        }
+
+        var subjectName = jsonObject.SelectToken("$.FullTable.SubjectMeta.SubjectName")?.ToObject<string>();
+        if (subjectName == null)
+        {
+            throw new InvalidOperationException("Permalink found with no SubjectName");
+        }
+
+        return new Permalink
+        {
+            Id = Guid.Parse(id),
+            PublicationTitle = publicationName,
+            DataSetTitle = subjectName,
+            ReleaseId = null,
```

Comment:
So there's no way we can retrieve the releaseId?  Perhaps using the `PublicationTitle` and `SubjectId` somehow?  Probably we couldn't infer this for every Release that has amendments, but possibly we could for a majority?

Reply:
The `ReleaseId` isn't used for anything currently, but I thought it would be a good idea to begin recording it for new permalinks generated from this point onwards.

Happy to remove it altogether if you don't agree with storing it, since it's not used.

I think we possibly could infer the release id from the subject id for the majority of permalinks, but there's going to be edge cases where we can't find a subject id in either private or public stats databases, and like you say, where the release has been amended we won't know which release it was created from.

I'm not sure it's worth doing this at the moment. Since nothing is using `ReleaseId` yet, we can back-fill them at any point in time.

Reply:
Yup agreed it's nothing that's needed for now, or perhaps ever!